File: App/app/src/main/java/com/cinema/ticket_booking/ui/booking/SelectSeatViewModel.java
package com.cinema.ticket_booking.ui.booking;

import androidx.lifecycle.*;
import com.cinema.ticket_booking.data.model.response.*;
import com.cinema.ticket_booking.data.repository.*;
import com.cinema.ticket_booking.util.Resource;
import java.util.*;
import dagger.hilt.android.lifecycle.HiltViewModel;
import javax.inject.Inject;

@HiltViewModel
public class SelectSeatViewModel extends ViewModel {
    private final ShowtimeRepository showtimeRepo;
    private final BookingRepository bookingRepo;

    private final MutableLiveData<Resource<SeatMapResponse>> seatMap = new MutableLiveData<>();
    private final MutableLiveData<SeatMapResponse> seatRefresh = new MutableLiveData<>();
    private final MutableLiveData<Resource<List<ComboResponse>>> combos = new MutableLiveData<>();
    private final Set<String> selectedSeatIds = new LinkedHashSet<>();
    private final Map<String, Integer> selectedCombos = new LinkedHashMap<>();
    private String currentShowtimeId;

    // Shared booking state
    public static List<String> pendingSeatIds = new ArrayList<>();
    public static Map<String, Integer> pendingCombos = new LinkedHashMap<>();
    public static double pendingTotalAmount;

    @Inject
    public SelectSeatViewModel(ShowtimeRepository showtimeRepo, BookingRepository bookingRepo) {
        this.showtimeRepo = showtimeRepo;
        this.bookingRepo = bookingRepo;
    }

    public LiveData<Resource<SeatMapResponse>> getSeatMap() {
        return seatMap;
    }

    public Set<String> getSelectedSeatIds() {
        return selectedSeatIds;
    }

    public Map<String, Integer> getSelectedCombos() {
        return selectedCombos;
    }

    public LiveData<SeatMapResponse> getSeatRefresh() {
        return seatRefresh;
    }

    public void loadSeatMap(String showtimeId) {
        this.currentShowtimeId = showtimeId;
        showtimeRepo.getSeatMap(showtimeId).observeForever(seatMap::setValue);
    }

    /** Silent background refresh — does not trigger loading indicator. */
    public void refreshSeatStatuses() {
        if (currentShowtimeId == null) return;
        showtimeRepo.getSeatMap(currentShowtimeId).observeForever(r -> {
            if (r != null && r.isSuccess() && r.data != null) {
                seatRefresh.postValue(r.data);
            }
        });
    }

    public boolean toggleSeat(SeatMapResponse.SeatItem seat) {
        if ("BOOKED".equals(seat.getStatus()) || "LOCKED".equals(seat.getStatus()))
            return false;
        if (selectedSeatIds.contains(seat.getShowtimeSeatId()))
            selectedSeatIds.remove(seat.getShowtimeSeatId());
        else
            selectedSeatIds.add(seat.getShowtimeSeatId());
        return true;
    }

    public double calculateTotal(SeatMapResponse seatMapData) {
        if (seatMapData == null)
            return 0;
        double total = 0;
        for (SeatMapResponse.SeatItem seat : seatMapData.getSeats()) {
            if (selectedSeatIds.contains(seat.getShowtimeSeatId()))
                total += seat.getPrice();
        }
        return total;
    }
}

File: App/app/src/main/java/com/cinema/ticket_booking/ui/booking/SelectSeatViewModel.kt
package com.cinema.ticket_booking.ui.booking

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.cinema.ticket_booking.data.model.response.ComboResponse
import com.cinema.ticket_booking.data.model.response.SeatMapResponse
import com.cinema.ticket_booking.data.repository.BookingRepository
import com.cinema.ticket_booking.data.repository.ShowtimeRepository
import com.cinema.ticket_booking.util.Resource
import dagger.hilt.android.lifecycle.HiltViewModel
import java.util.LinkedHashMap
import javax.inject.Inject

@HiltViewModel
class SelectSeatViewModel @Inject constructor(
    private val showtimeRepo: ShowtimeRepository,
    private val bookingRepo: BookingRepository
) : ViewModel() {

    private val _seatMap = MutableLiveData<Resource<SeatMapResponse>>()
    val seatMap: LiveData<Resource<SeatMapResponse>> = _seatMap

    private val _seatRefresh = MutableLiveData<SeatMapResponse>()
    val seatRefresh: LiveData<SeatMapResponse> = _seatRefresh

    private val _combos = MutableLiveData<Resource<List<ComboResponse>>>()
    val combos: LiveData<Resource<List<ComboResponse>>> = _combos

    val selectedSeatIds: MutableSet<String> = LinkedHashSet()
    val selectedCombos: MutableMap<String, Int> = LinkedHashMap()
    private var currentShowtimeId: String? = null

    companion object {
        @JvmField
        var pendingSeatIds: List<String> = ArrayList()
        @JvmField
        var pendingCombos: Map<String, Int> = LinkedHashMap()
        @JvmField
        var pendingTotalAmount: Double = 0.0
    }

    fun loadSeatMap(showtimeId: String) {
        currentShowtimeId = showtimeId
        showtimeRepo.getSeatMap(showtimeId).observeForever { _seatMap.value = it }
    }

    /** Silent background refresh — does not trigger loading indicator. */
    fun refreshSeatStatuses() {
        val showtimeId = currentShowtimeId ?: return
        showtimeRepo.getSeatMap(showtimeId).observeForever { r ->
            if (r != null && r.isSuccess && r.data != null) {
                _seatRefresh.postValue(r.data)
            }
        }
    }

    fun toggleSeat(seat: SeatMapResponse.SeatItem): Boolean {
        if ("BOOKED" == seat.status || "LOCKED" == seat.status) return false
        val seatId = seat.showtimeSeatId ?: return false
        if (selectedSeatIds.contains(seatId)) {
            selectedSeatIds.remove(seatId)
        } else {
            selectedSeatIds.add(seatId)
        }
        return true
    }

    fun calculateTotal(seatMapData: SeatMapResponse?): Double {
        if (seatMapData?.seats == null) return 0.0
        var total = 0.0
        for (seat in seatMapData.seats) {
            if (selectedSeatIds.contains(seat.showtimeSeatId)) {
                total += seat.price
            }
        }
        return total
    }
}
